src/collections/advanced_cache.c

```diff
@@ -0,0 +1,364 @@
+                _ze_advanced_cache_query_parse_time(&param.value, &params->time);
+            }
+        }
+    }
+}
+
+static bool _ze_advanced_cache_range_contains(const _ze_advanced_cache_range_t *range, uint32_t sn) {
+    return ((range->start == _ZE_ADVANCED_CACHE_QUERY_PARAMETERS_RANGE_UNBOUNDED || range->start <= sn) &&
+            (range->end == _ZE_ADVANCED_CACHE_QUERY_PARAMETERS_RANGE_UNBOUNDED || sn <= range->end));
+}
+
+static void _ze_advanced_cache_query_handler(z_loaned_query_t *query, void *ctx) {
+    _ze_advanced_cache_t *cache = (_ze_advanced_cache_t *)ctx;
+
+    z_view_string_t param_str;
+    z_query_parameters(query, &param_str);
+
+    _ze_advanced_cache_query_parameters_t params;
+    _ze_advanced_cache_query_parse_parameters(&params, z_view_string_loan(&param_str));
+
+    _z_time_since_epoch now;
+    z_result_t res = _z_get_time_since_epoch(&now);
+    if (res != _Z_RES_OK) {
+        _Z_ERROR("Dropped advanced cache query - failed to determine current system time: %i", res);
+        return;
+    }
+    _z_ntp64_t now_ntp64 = _z_timestamp_ntp64_from_time(now.secs, now.nanos);
+
+#if Z_FEATURE_MULTI_THREAD == 1
+    res = _z_mutex_lock(&cache->_outbox_mutex);
+    if (res != _Z_RES_OK) {
+        _Z_ERROR("Dropped advanced cache query - failed to lock outbox mutex: %i", res);
+        return;
+    }
+    res = _z_mutex_lock(&cache->_mutex);
+    if (res != _Z_RES_OK) {
+        _Z_ERROR("Dropped advanced cache query - failed to lock mutex: %i", res);
+        _z_mutex_unlock(&cache->_outbox_mutex);
+        return;
+    }
+#endif
+    size_t cap = _z_sample_ring_capacity(&cache->_cache);
+    size_t max = (params.max != _ZE_ADVANCED_CACHE_QUERY_PARAMETERS_MAX_UNBOUNDED) ? params.max : cap;
+    if (max > cap) max = cap;
+    if (max > cache->_outbox_cap) max = cache->_outbox_cap;
+
+    const bool range_filter = (params.range.start != _ZE_ADVANCED_CACHE_QUERY_PARAMETERS_RANGE_UNBOUNDED) ||
+                              (params.range.end != _ZE_ADVANCED_CACHE_QUERY_PARAMETERS_RANGE_UNBOUNDED);
+    const bool time_filter =
+        (params.time.start.bound != _Z_TIME_BOUND_UNBOUNDED) || (params.time.end.bound != _Z_TIME_BOUND_UNBOUNDED);
+
+    _z_sample_ring_reverse_iterator_t iter = _z_sample_ring_reverse_iterator_make(&cache->_cache);
+
+    size_t to_send = 0;
+    while (max > 0 && _z_sample_ring_reverse_iterator_next(&iter)) {
+        _z_sample_t *sample = _z_sample_ring_reverse_iterator_value(&iter);
+        if (range_filter && (!_z_source_info_check(&sample->source_info) ||
+                             !_ze_advanced_cache_range_contains(&params.range, sample->source_info._source_sn))) {
+            continue;
+        }
+        if (time_filter && (!_z_timestamp_check(&sample->timestamp) ||
+                            !_z_time_range_contains_at_time(&params.time, sample->timestamp.time, now_ntp64))) {
+            continue;
+        }
+
+        res = _z_sample_copy(&cache->_outbox[to_send], sample);
```

Comment:
If `_z_sample_copy` fails, the reply will have fewer than requested samples, but the loop continues. Can it be problem?